airbyte-cdk/bulk/toolkits/load-db/src/main/kotlin/io/airbyte/cdk/load/orchestration/db/direct_load_table/DirectLoadTableStreamLoader.kt

```diff
@@ -194,61 +194,77 @@ class DirectLoadTableDedupTruncateStreamLoader(
     }
 
     override suspend fun close(hadNonzeroRecords: Boolean, streamFailure: StreamProcessingFailed?) {
-        if (finalTableMaybeCorrectGeneration) {
-            val shouldUpsertUnconditionally: Boolean
-            if (initialStatus.realTable == null) {
-                // the real table doesn't yet exist.
-                // we should create it now, and upsert this sync's records into it.
-                // this won't cause downtime, because there was no data to begin with.
-                sqlTableOperations.createTable(
-                    stream,
-                    realTableName,
-                    columnNameMapping,
-                    replace = true,
-                )
-                shouldUpsertUnconditionally = true
-            } else if (
-                initialStatus.realTable.isEmpty ||
-                    nativeTableOperations.getGenerationId(realTableName) >=
-                        stream.minimumGenerationId
-            ) {
-                // the real table exists, but is empty.
-                // we should enforce its schema, and upsert this sync's records.
-                // again: this won't cause downtime, because there was no data to begin with.
-                nativeTableOperations.ensureSchemaMatches(stream, realTableName, columnNameMapping)
-                shouldUpsertUnconditionally = true
-            } else {
-                // the real table exists and has data. we should do more stringent checks
-                // before upserting, because we might need to truncate this table
-                // before upserting.
-                shouldUpsertUnconditionally = false
-            }
+        // Early return if there's a stream failure - no data operations needed
+        if (streamFailure != null) {
```

Comment:
I think this would be easier to follow: (i.e. no early returns) 
```kotlin
        if (streamFailure == null) {
            if (finalTableMaybeCorrectGeneration && shouldUpsertUnconditionally()) {
                // Direct upsert path for simpler cases
                performDirectUpsert()
            } else {
                // Needs temp table and overwrite approach
                performUpsertWithTemporaryTable()
            }
        }
```